plugins/janus_streaming.c

```diff
@@ -5052,41 +5057,98 @@ static int janus_streaming_rtsp_parse_sdp(const char *buffer, const char *name,
 			mcast = inet_addr(ip);
 		}
 	}
-	int port;
+
+	gboolean should_select_random_port = rtsp_client_port_range == NULL || strlen(rtsp_client_port_range) == 0;
+	uint16_t min_port = 0x400, max_port = 0xFFFF;
+	if(!should_select_random_port) {
+		/* Split in min and max port */
+		char *tmp_str = g_strdup(rtsp_client_port_range);
+		char *dash_pos = strrchr(tmp_str, '-');
+		if(dash_pos == NULL) {
+			min_port = atoi(tmp_str);
+		} else {
+			*dash_pos = '\0';
+			dash_pos++;
+			max_port = atoi(dash_pos);
+			min_port = atoi(tmp_str);
+		}
+		g_free(tmp_str);
+
+		if(min_port > max_port) {
+			uint16_t temp_port = min_port;
+			min_port = max_port;
+			max_port = temp_port;
+		}
+		if(min_port == 0)
+			min_port = 0x400;
+		if(max_port == 0)
+			max_port = 0xFFFF;
+		JANUS_LOG(LOG_INFO, "RTSP client port range: %u -- %u\n", min_port, max_port);
+	}
+	uint16_t port = should_select_random_port ? 0 : min_port;
 	struct sockaddr_in address;
 	socklen_t len = sizeof(address);
 	/* loop until can bind two adjacent ports for RTP and RTCP */
+	gboolean is_port_found;
 	do {
-		fds->fd = janus_streaming_create_fd(0, mcast, iface, media, media, name);
+		is_port_found = TRUE;
+		fds->fd = janus_streaming_create_fd(should_select_random_port ? 0 : port, mcast, iface, media, media, name);
 		if(fds->fd < 0) {
-			return -1;
-		}
-		if(getsockname(fds->fd, (struct sockaddr *)&address, &len) < 0) {
-			JANUS_LOG(LOG_ERR, "[%s] Bind failed for %s...\n", name, media);
-			close(fds->fd);
-			return -1;
+			is_port_found = FALSE;
+		} else {
+			if(getsockname(fds->fd, (struct sockaddr *)&address, &len) < 0) {
+				close(fds->fd);
+				if(should_select_random_port) {
+					JANUS_LOG(LOG_ERR, "[%s] Bind failed for %s...\n", name, media);
+					return -1;
+				}
+				is_port_found = FALSE;
+			} else {
+				if(should_select_random_port) {
+					uint16_t sin_port = ntohs(address.sin_port);
+					if(sin_port & 1) {
+						close(fds->fd);
+						is_port_found = FALSE;
+					} else {
+						port = sin_port;
+					}
+				} else if(port + 1 > max_port) {
+					close(fds->fd);
+					is_port_found = FALSE;
+				}
+				
+				if(is_port_found) {
+					fds->rtcp_fd = janus_streaming_create_fd(port+1, mcast, iface, media, media, name);
+					if(fds->rtcp_fd < 0) {
+						close(fds->fd);
+						is_port_found = FALSE;
+					} else if(getsockname(fds->rtcp_fd, (struct sockaddr *)&address, &len) < 0) {
+						close(fds->fd);
+						close(fds->rtcp_fd);
+						is_port_found = FALSE;
+					}
+				}
+			}
```

Comment:
This looks like a lot of code that could be greatly simplified. Check the `janus_nosip_allocate_port_pair` and `janus_nosip_bind_socket` methods in NoSIP for instance. Doing something similar would make it quite easy to do the same here, and not only for RTSP, but for all random ports in RTP mountpoints too (that is, any time you pass 0 there).

Reply:
thanks, will check